### ISSUE TITLE
pipe translate could not be found

### ISSUE THREAD

**Issue by saisaraswathi** [ngx-translate/core#566]
zone.js:388 Unhandled Promise rejection: Template parse errors:
The pipe 'translate' could not be found ("an>-->
					<select><option>Select language</option>
						<option (click)="changeLanguage('en')">[ERROR ->]{{ 'Languages.English' | translate }}</option>
							<option  (click)="changeLanguage('id')">{{ 'La"): DashboardComponent@12:45
The pipe 'translate' could not be found ("n')">{{ 'Languages.English' | translate }}</option>
							<option  (click)="changeLanguage('id')">[ERROR ->]{{ 'Languages.Bahasa' | translate }}</option></select>
	    		</div>
	    		<div class="col-md-offs"): DashboardComponent@13:47 ; Zone: <root> ; Task: Promise.then ; Value: Error: Template parse errors:
The pipe 'translate' could not be found ("an>-->
					<select><option>Select language</option>
						<option (click)="changeLanguage('en')">[ERROR ->]{{ 'Languages.English' | translate }}</option>
							<option  (click)="changeLanguage('id')">{{ 'La"): DashboardComponent@12:45
The pipe 'translate' could not be found ("n')">{{ 'Languages.English' | translate }}</option>
							<option  (click)="changeLanguage('id')">[ERROR ->]{{ 'Languages.Bahasa' | translate }}</option></select>
	    		</div>
	    		<div class="col-md-offs"): DashboardComponent@13:47

**Comment by saisaraswathi** [comment]
app.component.ts:


import { Component } from '@angular/core';
import { TranslateService } from 'ng2-translate';
// import {DashboardComponent} from './dashboard/dashboard.component';
import {Subscription } from 'rxjs';
import {Router, ActivatedRoute} from '@angular/router';
import {OnInit, OnDestroy} from '@angular/core';

@Component({
    selector: 'pm-app',
    templateUrl : './app/app.component.html'
})
export class AppComponent { 
   private subscription: Subscription;
  

  constructor(private translate: TranslateService, private activatedRoute: ActivatedRoute) {
        translate.addLangs(["en", "id"]);
        translate.setDefaultLang('en');

        let browserLang = translate.getBrowserLang();
        translate.use(browserLang.match(/en|id/) ? browserLang : 'en');
    }

    changeLanguage(lang: any){
        this.translate.use(lang);
    }

    ngOnInit() {
    // subscribe to router event
    this.subscription = this.activatedRoute.queryParams.subscribe(
      (param: any) => {
        let locale = param['locale'];
        if (locale !== undefined){
            this.translate.use(locale);
        }
      });
  }

  ngOnDestroy() {
    // prevent memory leak by unsubscribing
    this.subscription.unsubscribe();
  }
}

**Comment by saisaraswathi** [comment]
can anyone plz help me to resolve this issue?

**Comment by SamVerschueren** [comment]
Make sure to import the `TranslateModule` as described [here](https://github.com/ngx-translate/core#1-import-the-translatemodule). If that doesn't work, try setting up a plunkr to reproduce this. Just throwing in component code isn't really helpful as this is most likely a problem with configuring the module.

Also, there is an issue template for a reason. Because it helps us to identify the problem faster.

**Comment by saisaraswathi** [comment]
app.module.ts

import { NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { FormsModule } from '@angular/forms';
import { HttpModule, Http } from '@angular/http';
import {RegistrationComponent} from './dashboard/user_registration/user_registration.component';
import { AppComponent }  from './app.component';
import { routing } from './app.routing';
import { TranslateService } from 'ng2-translate';
import { TranslateModule, MissingTranslationHandler } from 'ng2-translate';

import {DashboardComponent} from './dashboard/dashboard.component';
import {DeviceComponent} from './dashboard/device/device.component';
import {ResetDeviceComponent} from './dashboard/device/resetDevice/ResetDevice.component';
import {LoginComponent} from './login/login.component';
import { AuthenticationService} from './services/authentication.service';
import { DeviceService } from './services/device.service';
import { AlertService } from './services/alert.service';
import { MyMissingTranslationHandler } from './missingtemplate.component';


@NgModule({
  imports: [ BrowserModule, 
             FormsModule,
             HttpModule,
             routing,
                TranslateModule.forRoot()
             ],
  declarations: [ AppComponent ,
                  LoginComponent,
                  DashboardComponent,
                  DeviceComponent,RegistrationComponent,
                  ResetDeviceComponent],
  providers: [ AuthenticationService,
               DeviceService,
               AlertService,{ provide: MissingTranslationHandler, useClass: MyMissingTranslationHandler}
              ],
  exports: [
        TranslateModule
    ],
  bootstrap: [ AppComponent ]
})
export class AppModule { }


I imported translate module, can u plz check now?

**Comment by SamVerschueren** [comment]
I don't see any reason why it shouldn't work. The only way we can debug this decently is if you provide us with a git repository or a plunkr with a minimal application which reproduces this behaviour.

**Comment by SamVerschueren** [comment]
I also noticed that you still run ng2-translate, I'd recommend upgrading to ngx-translate.

**Comment by saisaraswathi** [comment]
i am behind corporate proxy unable to upload into github and angular2 files are many so not in plunker. Is there any other to reproduce the code?

**Comment by saisaraswathi** [comment]
i upgraded to ngx-translate

**Comment by SamVerschueren** [comment]
Here's a basic plunkr to start from. https://plnkr.co/edit/01UjWY3TKfP6pgwXKuEa?p=preview

**Comment by saisaraswathi** [comment]
i did exactly what they did in plunker example, still it is saying "The pipe 'translate' could not be found "

**Comment by phattranky** [comment]
I got this problem before, Then i try to import TranslateModule in childre. It's working

**Comment by VickNet** [comment]
in my case I have a submodule. I added TranslateModule in my imports and it works.
Thanks @phattranky 👍

**Comment by rajjaiswalsaumya** [comment]
Can someone post full code, i am getting same error

**Comment by bigsnowdrop** [comment]
solved the problem so
add "TranslateModule" to exports
`import { NgModule } from '@angular/core';
import { HttpClientModule, HttpClient } from '@angular/common/http';
import { Select2Module } from 'ng2-select2';
import { TranslateModule, TranslateLoader } from '@ngx-translate/core';
import { AppTranslationService, TranslateLanguageLoader } from './services';

@NgModule({
    declarations: [
    ],
    imports: [
        Select2Module,
        HttpClientModule,
        TranslateModule.forRoot({
            loader: {
                provide: TranslateLoader,
                useClass: TranslateLanguageLoader
            }
        })
    ],
    exports: [
        TranslateModule
    ],
    providers: [
        AppTranslationService
    ]
})
export class SharedModule {}`

**Comment by bigbharatjain** [comment]
**app.module.ts**
```
import { TranslateModule, TranslateLoader } from '@ngx-translate/core';
import { TranslateHttpLoader } from '@ngx-translate/http-loader';
import { HttpClientModule, HttpClient } from '@angular/common/http';
export function createTranslateLoader(http: HttpClient) {
    return new TranslateHttpLoader(http, './assets/i18n/', '.json');
}
@NgModule({
    declarations: [
        AppComponent
    ],
    imports: [
        BrowserModule,
        FormsModule,
        HttpModule,
        NgbModule.forRoot(),
        CoreModule,
        AppRoutingModule,
        ReactiveFormsModule,
        BrowserAnimationsModule,
        SharedModule,
        HttpClientModule,
        TranslateModule.forRoot({
            loader: {
                provide: TranslateLoader,
                useFactory: createTranslateLoader,
                deps: [HttpClient]
            }
        })
    ],

    providers: [
        {provide: BrowserXhr, useClass: BrowserCredentials},
        {provide: RequestOptions, useClass: GlobalHttpOptions},
        Title
    ],
    bootstrap: [AppComponent],
    exports: []

})
```
**app.component.ts**
```
import {Component, OnInit} from '@angular/core';

import {Title} from '@angular/platform-browser';
import {ConfigurationService} from './shared/services/configuration.service';
import { TranslateService } from '@ngx-translate/core';
@Component({
    selector: 'cc-root',
    template: `
        <cc-authentication-loader></cc-authentication-loader>
        <router-outlet></router-outlet>`,
    styles: []
})
export class AppComponent implements OnInit {
    constructor(private title: Title, translate: TranslateService) {
        translate.setDefaultLang('en');
    }

    ngOnInit(): void {
        this.title.setTitle('Carcare');
    }
}

```
In other component, when I add 
`<h1>{{"Welcome"|translate}}</h1>` 
It gives me the error 
`The pipe 'translate' could not be found`

I am using angular 5. 
Please help what I am missing ?

**Comment by webeagle199066** [comment]
Привет
подключи **TranslateModule** в нужном модуле, где вызываешь **| translate**

dashboard.module.ts
`@NgModule({
    imports: [
        CommonModule,
        RouterModule.forChild([{
            path: '',
            component: DashboardComponent
        }]),
        TranslateModule
    ],
    declarations: [
        DashboardComponent
    ],
    providers: [
        {provide: MODULE_CONFIG, useValue: moduleConfig}
    ]
})`

dashboard.component.html
`{{ 'dashboard.workStart.text' | translate }}`

И ни каких ошибок =)

**Comment by HonzoBonzo** [comment]
Had the same issue, as described above. Couldn't find pipe translate... It happens with the newest versions:     

"@ngx-translate/core": "^10.0.0",
"@ngx-translate/http-loader": "^3.0.1",

Downgrading to:
"@ngx-translate/core": "^9.1.1",
"@ngx-translate/http-loader": "^2.0.1",

solved this problem.

AppModule:
```
export function createTranslateLoader(http: HttpClient) {
  return new TranslateHttpLoader(http, './assets/i18n/', '.json');
}
```


```
@NgModule({
imports: [    TranslateModule.forRoot({
      loader: {
        provide: TranslateLoader,
        useFactory: createTranslateLoader,
        deps: [HttpClient]
      }
    })]
...
```

good luck

**Comment by CodeAndWeb** [comment]
ngx-translate requires Angular 6.
You have to use ngx-translate 9.x with Angular 5.

**Comment by ishwaryanair** [comment]
i did downgrading to:
"@ngx-translate/core": "^9.1.1",
"@ngx-translate/http-loader": "^2.0.1",

and import into TranslateModule into submodule is also done. Still i dont see translated text shown.

**Comment by ishwaryanair** [comment]
i did downgrading to:
"@ngx-translate/core": "^9.1.1",
"@ngx-translate/http-loader": "^2.0.1",

and import into TranslateModule into submodule is also done. Still i dont see translated text shown.

**Comment by CodeAndWeb** [comment]
Can you put your project on github?

**Comment by ishwaryanair** [comment]
I got it working by using {{translate.instant('my text')}} but direct translate directive within html tag dint work for me.

**Comment by bleuscyther** [comment]
**ionic/angular** 
For anyone that may find this helpful : sometimes  you have to kill the ionic process and run it again. Happens on ionic serve if you install some new package

**Comment by tiberiumihai** [comment]
Any updates on this? 

I got it working like @ishwaryanair said but the HTML looks messy now because the pipe is not used.

`<p [innerHTML]="translate.instant(model.title)"></p>` .  <--- WORKS

`<p [innerHTML]="model.title | translate"></p>` .  <--- DOESN'T WORK: The pipe 'translate' could not be found


Using:
```
"@angular/core": "^6.1.3",
"@ngx-translate/core": "^10.0.2",
"@ngx-translate/http-loader": "^3.0.1",
"typescript": "^2.9.2",
```

**Comment by BruneXX** [comment]
Hey guys I've followed the documentation instructions but using translate in other component than app.component is causing this error, the only solution is to make a downgrade? thanks

**Comment by BruneXX** [comment]
FYI: I've found a solution to this, take a look at this: https://github.com/ngx-translate/core/issues/566#issuecomment-323280513

**Comment by erperejildo** [comment]
@BruneXX is going to be used everywhere. Do we really need to modify every single module? Can't it be used as shared module?

**Comment by erperejildo** [comment]
Can you share it? I created this shared module and I added there my translation module and I can add different providers or other modules but I need to add this in every module anyway. I tried adding this on `app.module` but doesn't work

**Comment by BruneXX** [comment]
Hi @erperejildo That's the way angular works, take a look at this: https://angular.io/guide/sharing-ngmodules
AFAIK: You must add your SharedModule in modules that you need to use "translate" or "any" other module/component.

Like @tiberiumihai show in his comment https://github.com/ngx-translate/core/issues/566#issuecomment-414062997

**Comment by erperejildo** [comment]
Thanks I got it working yesterday

**Comment by tiberiumihai** [comment]
@erperejildo Than your SharedModule will look something like this, adding your providers in forRoot().providers array:

shared.module.ts
```
import { NgModule, ModuleWithProviders } from '@angular/core';
import { TranslateModule } from '@ngx-translate/core';
import { FormsModule } from '@angular/forms';

import { MyProvider } from "./services/my-provider.service.ts";

@NgModule({
  imports: [
    TranslateModule,
    FormsModule
  ],
  exports: [
    TranslateModule,
    FormsModule
  ]
})
export class SharedModule {

  static forRoot(): ModuleWithProviders {
    return {
      ngModule: SharedModule,
      providers: [
        MyProvider
      ]
    }
  }
}
```

**Comment by erperejildo** [comment]
@tiberiumihai where do you move this (more info https://ionicframework.com/docs/developer-resources/ng2-translate/#lazy-loading-pages):
```
TranslateModule.forChild({
      loader: {
        provide: TranslateLoader,
        useFactory: (TranslateLoaderFactory),
        deps: [HttpClient]
      }
    })
```

Also, when they say **"DO NOT include providers inside shared modules"** they only mean inside  `@NgModule`? This doesn't apply to the shared class?

**Comment by indianazhao** [comment]
I solved this issue by exporting 'TranslatePipe' in the ShareModule.

```typescript
// AoT requires an exported function for factories
export function createTranslateLoader(http: HttpClient) {
  return new TranslateHttpLoader(http, '../assets/i18n/', '.json');
}

@NgModule({
  imports: [
    CommonModule,
    TranslateModule.forRoot({
      loader: {
        provide: TranslateLoader,
        useFactory: (createTranslateLoader),
        deps: [HttpClient]
      }
    }),
  ],
  declarations: [],
  exports: [
    FormsModule,
    TranslateModule,
    TranslatePipe, // <-- Export the pipe
  ]
})
export class ShareModule {

  constructor(private translate: TranslateService) {
    // this language will be used as a fallback when a translation isn't found in the current language
    translate.setDefaultLang('zh-tw');

    // the lang to use, if the lang isn't available, it will use the current loader to get them
    translate.use('zh-tw');
  }
}
```

**Comment by matheusmurta** [comment]
For problemns with Jasmine test.

Its works for me guys 😉

Add this on MyComponent..component.spec.ts
import { TranslateModule } from "@ngx-translate/core";


 beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [ EmailMessageComponent ],
      imports: [
        TranslateModule.forRoot(),
      ]
    })
    .compileComponents();
  }));

**Comment by artworkjpm** [comment]
what worked for me was adding **TranslateModule** to` shared.module` file:

```

import { TranslateModule } from "@ngx-translate/core";

@NgModule({
  imports: [TranslateModule, ],
  
  exports: [TranslateModule],
  providers: [],
})
export class SharedModule {}

```

**Comment by yoedusvany** [comment]
Hi guys, I have the same issue, 
Angular 9.1.7
I downgrownded this packages...
"@ngx-translate/core": "^12.1.2",
"@ngx-translate/http-loader": "^4.0.0",

Before I had  
"@ngx-translate/core": "^13.0.0",
"@ngx-translate/http-loader": "^6.0.0",

This is my app.module.ts
....`
import { TranslateLoader, TranslateModule } from '@ngx-translate/core';
import { TranslateHttpLoader } from '@ngx-translate/http-loader';
imports: [
...
 TranslateModule.forRoot({
            loader: {provide: TranslateLoader, useFactory: TranslationLoaderFactory, deps: [HttpClient]}
        })
]
`
My app.component.ts
`import { environment } from './../environments/environment';
import {TranslateService} from '@ngx-translate/core';
import { Component, Inject, PLATFORM_ID, Optional } from '@angular/core';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.scss']
})
export class AppComponent{
  title = 'falcon';

  constructor(private translateService: TranslateService) {
    this.translateService.addLangs(['es', 'en']);
    this.translateService.setDefaultLang(environment.defaultLanguage);
    this.translateService.use(environment.defaultLanguage);

    this.translateService.get('VALIDATIONS.REQUIRED').subscribe((res: string) => {
        console.log(res);
    });
  }

  public useLanguage(lang: string): void {
    this.translateService.setDefaultLang(lang);
  }

}
`
And the service is working!!!! but the pipe throw an exception

Any Idea?

**Comment by jboucly** [comment]
If you use `shared.module` made as shown @artworkjpm ! 
And if not work once again I just had to restart my local server and voila :) 
Thanks 🔥

**Comment by SpenceDiNicolantonio** [comment]
TLDR: If anyone is having this problem while using a `SharedModule`, make sure `SharedModule` is **NOT** imported in `AppModule`

I couldn't figure out for the life of me why this wouldn't work (in only one component) and I verified that my `SharedModule` and submodule imports/exports were setup properly. @tiberiumihai's [comment](https://github.com/ngx-translate/core/issues/566#issuecomment-414062997) lead me to discover that `SharedModule` was being erroneously imported into the main `AppModule`. Removing that fixed everything!
